File: xNet/src/test/java/org/sqlunet/provider/QueriesUnitTest.java
package org.sqlunet.provider;

import org.junit.Test;
import org.sqlunet.provider.XSqlUNetDispatcher.Result;

import java.util.Arrays;

public class QueriesUnitTest
{
	private final int[] codes = {XSqlUNetDispatcher.PREDICATEMATRIX, XSqlUNetDispatcher.PREDICATEMATRIX_VERBNET, XSqlUNetDispatcher.PREDICATEMATRIX_PROPBANK, XSqlUNetDispatcher.PREDICATEMATRIX_FRAMENET, XSqlUNetDispatcher.WORDS_FNWORDS_PBWORDS_VNWORDS, XSqlUNetDispatcher.WORDS_PBWORDS_VNWORDS, XSqlUNetDispatcher.WORDS_VNWORDS_VNCLASSES, XSqlUNetDispatcher.WORDS_VNWORDS_VNCLASSES_U, XSqlUNetDispatcher.WORDS_PBWORDS_PBROLESETS, XSqlUNetDispatcher.WORDS_PBWORDS_PBROLESETS_U, XSqlUNetDispatcher.WORDS_FNWORDS_FNFRAMES_U, XSqlUNetDispatcher.SOURCES,};
	private final String uriLast = "LAST";
	private final String[] projection = {"PROJ1", "PROJ2", "PROJ3"};
	private final String selection = "SEL";
	private final String[] selectionArgs = {"ARG1", "ARG2", "ARG3"};
	private final String sortOrder = "SORT";

	@Test
	public void queriesUnion()
	{
		for (int code : new int[]{XSqlUNetDispatcher.WORDS_VNWORDS_VNCLASSES_U, XSqlUNetDispatcher.WORDS_PBWORDS_PBROLESETS_U, XSqlUNetDispatcher.WORDS_FNWORDS_FNFRAMES_U})
		{
			System.out.println("CODE: " + code);
			Result r = QueriesLegacy.queryLegacy(code, uriLast, projection, selection, selectionArgs);
			System.out.println("table :" + r.table);
			System.out.println("projection :" + Arrays.toString(r.projection));
			System.out.println("selection :" + r.selection);
			System.out.println("args :" + Arrays.toString(r.selectionArgs));
			System.out.println("groupby :" + r.groupBy);
			System.out.println();
		}
	}

	@Test
	public void queriesLegacyAgainstProvider()
	{
		for (int i = 0; i < codes.length; i++)
		{
			int code = codes[i];
			queriesLegacyAgainstProvider(code, uriLast, projection, selection, selectionArgs, sortOrder);
		}
	}

	private void queriesLegacyAgainstProvider(final int code, final String uriLast, final String[] projection, final String selection, final String[] selectionArgs, final String sortOrder)
	{
		Result r1 = QueriesLegacy.queryLegacy(code, uriLast, projection, selection, selectionArgs);
		Result r2 = queryProvider(code, uriLast, projection, selection, selectionArgs);
		check(code, r1, r2);
	}

	public static Result queryProvider(final int code, final String uriLast, final String[] projection0, final String selection0, final String[] selectionArgs0)
	{
		Result r = queryProviderMain(code, uriLast, projection0, selection0, selectionArgs0);
		return r;
	}

	public static Result queryProviderMain(final int code, final String uriLast, final String[] projection0, final String selection0, final String[] selectionArgs0)
	{
		return XSqlUNetDispatcher.queryMain(code, uriLast, projection0, selection0, selectionArgs0);
	}

	private void check(final int code, final Result r1, final Result r2)
	{
		assert equals(r1.table, r2.table) : "Code=" + code + "\n" + r1.table + "\n!=\n" + r2.table;
		assert Arrays.equals(r1.projection, r2.projection) : "Code=" + code + "\n" + Arrays.toString(r1.projection) + "\n!=\n" + Arrays.toString(r2.projection);
		assert equals(r1.selection, r2.selection) : "Code=" + code + "\n" + r1.selection + "\n!=\n" + r2.selection;
		assert Arrays.equals(r1.selectionArgs, r2.selectionArgs) : "Code=" + code + "\n" + Arrays.toString(r1.selectionArgs) + "\n!=\n" + Arrays.toString(r2.selectionArgs);
		assert equals(r1.groupBy, r2.groupBy) : "Code=" + code + "\n" + r1.groupBy + "\n!=\n" + r2.groupBy;
	}

	private static boolean equals(Object a, Object b)
	{
		return (a == b) || (a != null && a.equals(b));
	}
}

File: xNet/src/test/java/org/sqlunet/provider/QueriesUnitTest.kt
/*
 * Copyright (c) 2023. Bernard Bou
 */
package org.sqlunet.provider

import org.junit.Test
import org.sqlunet.provider.XNetControl.queryMain

class QueriesUnitTest {

    private val codes = intArrayOf(
        XNetControl.PREDICATEMATRIX,
        XNetControl.PREDICATEMATRIX_VERBNET,
        XNetControl.PREDICATEMATRIX_PROPBANK,
        XNetControl.PREDICATEMATRIX_FRAMENET,
        XNetControl.WORDS_FNWORDS_PBWORDS_VNWORDS,
        XNetControl.WORDS_PBWORDS_VNWORDS,
        XNetControl.WORDS_VNWORDS_VNCLASSES,
        XNetControl.WORDS_VNWORDS_VNCLASSES_U,
        XNetControl.WORDS_PBWORDS_PBROLESETS,
        XNetControl.WORDS_PBWORDS_PBROLESETS_U,
        XNetControl.WORDS_FNWORDS_FNFRAMES_U,
        XNetControl.SOURCES,
        XNetControl.WORDS_VNWORDS_VNCLASSES_1,
        XNetControl.WORDS_VNWORDS_VNCLASSES_2,
        XNetControl.WORDS_VNWORDS_VNCLASSES_1U2,
        XNetControl.WORDS_VNWORDS_VNCLASSES_U,
        XNetControl.WORDS_PBWORDS_PBROLESETS_1,
        XNetControl.WORDS_PBWORDS_PBROLESETS_2,
        XNetControl.WORDS_PBWORDS_PBROLESETS_1U2,
        XNetControl.WORDS_PBWORDS_PBROLESETS_U,
        XNetControl.WORDS_FNWORDS_FNFRAMES_1,
        XNetControl.WORDS_FNWORDS_FNFRAMES_2,
        XNetControl.WORDS_FNWORDS_FNFRAMES_1U2,
        XNetControl.WORDS_FNWORDS_FNFRAMES_U
    )
    private val uriLast = "LAST"
    private val projection = arrayOf("PROJ1", "PROJ2", "PROJ3")
    private val selection = "SEL"
    private val selectionArgs = arrayOf("ARG1", "ARG2", "ARG3")
    private val sortOrder = "SORT"

    @Test
    fun queriesUnion() {
        for (code in intArrayOf(XNetControl.WORDS_VNWORDS_VNCLASSES_U, XNetControl.WORDS_PBWORDS_PBROLESETS_U, XNetControl.WORDS_FNWORDS_FNFRAMES_U)) {
            println("CODE: $code")
            val r = QueriesLegacy.queryLegacy(code, uriLast, projection, selection, selectionArgs)!!
            println("table :" + r.table)
            println("projection :" + r.projection.contentToString())
            println("selection :" + r.selection)
            println("args :" + r.selectionArgs.contentToString())
            println("groupby :" + r.groupBy)
            println()
        }
    }

    @Test
    fun queriesUnionLegacyAgainstProvider() {
        for (code in intArrayOf(XNetControl.WORDS_VNWORDS_VNCLASSES_U, XNetControl.WORDS_PBWORDS_PBROLESETS_U, XNetControl.WORDS_FNWORDS_FNFRAMES_U)) {
            queriesLegacyAgainstProvider(code, uriLast, projection, selection, selectionArgs, sortOrder)
        }
    }

    @Test
    fun queriesLegacyAgainstProvider() {
        for (code in codes) {
            queriesLegacyAgainstProvider(code, uriLast, projection, selection, selectionArgs, sortOrder)
        }
    }

    private fun queriesLegacyAgainstProvider(code: Int, uriLast: String, projection: Array<String>, selection: String, selectionArgs: Array<String>, @Suppress("unused") sortOrder: String) {
        val r1 = QueriesLegacy.queryLegacy(code, uriLast, projection, selection, selectionArgs)
        val r2 = queryProvider(code, uriLast, projection, selection, selectionArgs)
        check(code, r1, r2)
    }

    private fun check(code: Int, r1: XNetControl.Result?, r2: XNetControl.Result?) {
        assert(r1 != null)
        assert(r2 != null)
        assert(equals(r1!!.table, r2!!.table)) {
            """
            Code=$code table 
            ${r1.table}
            !=
            ${r2.table}
            """
        }
        assert(r1.projection.contentEquals(r2.projection)) {
            """
            Code=$code projection 
            ${r1.projection.contentToString()}
            !=
            ${r2.projection.contentToString()}
            """
        }
        assert(equals(r1.selection, r2.selection)) {
            """
            Code=$code selection 
            ${r1.selection}
            !=
            ${r2.selection}
            """
        }
        assert(r1.selectionArgs.contentEquals(r2.selectionArgs)) {
            """
            Code=$code args 
            ${r1.selectionArgs.contentToString()}
            !=
            ${r2.selectionArgs.contentToString()}
            """
        }
        assert(equals(r1.groupBy, r2.groupBy)) {
            """
            Code=$code group by 
            ${r1.groupBy}
            !=
            ${r2.groupBy}
            """
        }
    }

    companion object {

        fun queryProvider(code: Int, uriLast: String?, projection0: Array<String>?, selection0: String?, selectionArgs0: Array<String>?): XNetControl.Result? {
            return queryProviderMain(code, uriLast, projection0, selection0, selectionArgs0)
        }

        private fun queryProviderMain(code: Int, uriLast: String?, projection0: Array<String>?, selection0: String?, selectionArgs0: Array<String>?): XNetControl.Result? {
            return queryMain(code, projection0, selection0, selectionArgs0)
        }

        private fun equals(a: Any?, b: Any?): Boolean {
            return a == b
        }
    }
}
